service/policy/db/migrations/20250703000000_obligations.sql

```diff
@@ -0,0 +1,116 @@
+-- +goose Up
+-- +goose StatementBegin
+
+CREATE TABLE IF NOT EXISTS obligation_definitions
+(
+    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+    namespace_id UUID NOT NULL REFERENCES attribute_namespaces(id),
+    -- name is a unique identifier for the obligation definition within the namespace
+    name VARCHAR NOT NULL,
+    -- implicit index on unique (namespace_id, name) combo
+    -- index name: obligation_definitions_namespace_id_name_key
+    UNIQUE (namespace_id, name)
+);
+
+CREATE TABLE IF NOT EXISTS obligation_values_standard
+(
+    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+    obligation_definition_id UUID NOT NULL REFERENCES obligation_definitions(id),
+    -- value is a unique identifier for the obligation value within the definition
+    value VARCHAR NOT NULL,
+    -- implicit index on unique (obligation_definition_id, value) combo
+    -- index name: obligation_values_standard_obligation_definition_id_value_key
+    UNIQUE (obligation_definition_id, value)
+);
+
+CREATE TABLE IF NOT EXISTS obligation_triggers
+(
+    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+    attribute_value_id UUID NOT NULL REFERENCES attribute_values(id),
+    obligation_value_id UUID NOT NULL REFERENCES obligation_values_standard(id),
+    UNIQUE (attribute_value_id, obligation_value_id)
+);
+
+CREATE TABLE IF NOT EXISTS obligation_fulfillers
+(
+    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+    obligation_value_id UUID NOT NULL REFERENCES obligation_values_standard(id),
+    conditionals JSONB
+);
+
+CREATE TABLE IF NOT EXISTS obligation_action_attribute_values
+(
+    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+    obligation_value_id UUID NOT NULL REFERENCES obligation_values_standard(id),
+    action_id UUID NOT NULL REFERENCES actions(id),
+    attribute_value_id UUID NOT NULL REFERENCES attribute_values(id),
+    UNIQUE(obligation_value_id, action_id, attribute_value_id)
+);
+
+CREATE OR REPLACE FUNCTION get_obligation_tables()
```

Comment:
I think I prefer adding the metadata columns and triggers directly during table creation for readability, but I'm not fully opposed to this approach if you prefer it. However, if we do stick with this function approach for DRYness, we should clean up these functions at the end of the migration `up` instead of during the `down` as they're only utilized within the migration itself and not during regular queries/SQL statements.

Reply:
> Meanwhile, do you mind doing research on the reusable user-defined types if you feel strongly about that approach?

This is how you would do it FWIW using Postgres [DOMAIN](https://www.postgresql.org/docs/current/sql-createdomain.html)s per those docs and Virgil:

```postgresql
CREATE DOMAIN otdf_metadata AS JSONB;
CREATE DOMAIN otdf_timestamp AS TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP NOT NULL;

CREATE TABLE IF NOT EXISTS obligation_definitions
(
    -- ...other rows
    metadata otdf_metadata,
    created_at otdf_timestamp,
    updated_at otdf_timestamp
    -- ...constraints
);
```

Reply:
removed functions